csharp/src/Microsoft.ML.OnnxRuntime/Training/CheckpointState.shared.cs

```diff
@@ -47,19 +35,129 @@ public CheckpointState(string checkpointPath)
         /// Loads Checkpoint state from path
         /// </summary>
         /// <param name="checkpointPath"> absolute path to checkpoint</param>
-        private void LoadCheckpoint(string checkpointPath)
+        public static CheckpointState LoadCheckpoint(string checkpointPath)
         {
-            NativeApiStatus.VerifySuccess(NativeTrainingMethods.OrtLoadCheckpoint(NativeOnnxValueHelper.GetPlatformSerializedString(checkpointPath), out handle));
+            if (!NativeTrainingMethods.TrainingEnabled())
+            {
+                throw new InvalidOperationException("Training is disabled in the current build. Please build ONNXRuntime from source with the build flag enable_training_apis.\n");
+            }
+
+            var envHandle = OrtEnv.Instance().Handle; // just so it is initialized
+            IntPtr checkpointHandle = IntPtr.Zero;
+            NativeApiStatus.VerifySuccess(NativeTrainingMethods.OrtLoadCheckpoint(NativeOnnxValueHelper.GetPlatformSerializedString(checkpointPath), out checkpointHandle));
+
+            return new CheckpointState(checkpointHandle);
         }
 
         /// <summary>
         /// Saves the checkpoint
         /// <param name="checkpointPath"> absolute path to the checkpoint file.</param>
         /// <param name="includeOptimizerState"> absolute path to the checkpoint file.</param>
         /// </summary>
-        public void SaveCheckpoint(string checkpointPath, bool includeOptimizerState = false)
+        public static void SaveCheckpoint(CheckpointState state, string checkpointPath, bool includeOptimizerState = false)
+        {
+            NativeApiStatus.VerifySuccess(NativeTrainingMethods.OrtSaveCheckpoint(state.Handle, NativeOnnxValueHelper.GetPlatformSerializedString(checkpointPath), includeOptimizerState));
+        }
+
+        /// <summary>
+        /// Adds the given int property to the checkpoint state.
+        /// <param name="propertyName">Unique name of the property being added.</param>
+        /// <param name="propertyValue">Property value associated with the given name.</param>
+        /// </summary>
+        public void AddProperty(string propertyName, long propertyValue)
+        {
+            var propertyNameUtf8 = NativeOnnxValueHelper.StringToZeroTerminatedUtf8(propertyName);
+            long[] value = new long[1];
+            value[0] = propertyValue;
+            Memory<long> memory = value;
+            var memHandle = memory.Pin();
+            try
+            {
+                IntPtr memPtr;
+                unsafe
+                {
+                    memPtr = (IntPtr)memHandle.Pointer;
+                }
+                NativeApiStatus.VerifySuccess(NativeTrainingMethods.OrtAddProperty(handle, propertyNameUtf8, (long)0, memPtr));
+            }
+            catch (Exception)
```

Comment:
the `using` statement can automatically call Dispose at the end of a scope
https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/statements/using

e.g.
```c#
using (var memHandle = memory.Pin()) {
    ...
}
```